### PR TITLE
AssemblyScript SDK

### PR DESCRIPTION
**Pull request by RyanLassigBanks**
Adds AssemblyScript SDK for writing sabre smart contracts and sample contract

### Test

1) Build wasm
```
cd contracts/inc-dec-key
npm run asbuild
```

2) Start sabre environment
`docker-compose -f docker-compose-installed.yaml up --build`

3) Exec into sabre-cli container
`docker exec -it sabre-cli bash`

4) Upload and configure inc-dec-key contract
```
sabre cr --create incdec --owner $(cat root/.sawtooth/keys/root.pub) --url http://rest-api:9708
sabre upload --filename project/contracts/inc-dec-key/manifest.yaml --url http://rest-api:9708
sabre ns --create 12cd3c --url http://rest-api:9708 --owner $(cat /root/.sawtooth/keys/root.pub)
sabre perm 12cd3c incdec --read --write --url http://rest-api:9708
```

5) Run commands
```
cd projects/contracts/inc-dec-key/payloads

// initiate key
sabre exec --contract incdec:1.0 --payload  ./init --inputs 12cd3c  --outputs 12cd3c --url http://rest-api:9708

// increment key value by one
sabre exec --contract incdec:1.0 --payload  ./inc --inputs 12cd3c  --outputs 12cd3c --url http://rest-api:9708

// decrement key value by one
sabre exec --contract incdec:1.0 --payload  ./dec --inputs 12cd3c  --outputs 12cd3c --url http://rest-api:9708

// remove key
sabre exec --contract incdec:1.0 --payload  ./remove --inputs 12cd3c  --outputs 12cd3c --url http://rest-api:9708
```

### CODE REVIEW

**Comment by peterschwarz** [comment]
The above testing infrastructure works, but if I run a the `./dec` payload after deleting the key, I get the following:
```
sawtooth-sabre-tp  | 2020-01-29 21:47:24,978 INFO  [sawtooth_sabre::wasm_executor::wasm_module] Payload written to memory
sawtooth-sabre-tp  | 2020-01-29 21:47:24,978 INFO  [sawtooth_sabre::wasm_executor::wasm_module] Signer written to memory
sawtooth-sabre-tp  | 2020-01-29 21:47:24,978 INFO  [sawtooth_sabre::wasm_executor::wasm_module] Signature written to memory
sawtooth-sabre-tp  | 2020-01-29 21:47:24,980 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] ALLOC Execution time: 0 secs 0 ms
sawtooth-sabre-tp  | 2020-01-29 21:47:24,981 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] 105,110,99,44,116,101,115,116,48,10
sawtooth-sabre-tp  | 2020-01-29 21:47:24,982 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] ALLOC Execution time: 0 secs 0 ms
sawtooth-sabre-tp  | 2020-01-29 21:47:24,982 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] Payload: inc,test0
sawtooth-sabre-tp  |
sawtooth-sabre-tp  | 2020-01-29 21:47:25,012 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] ALLOC Execution time: 0 secs 0 ms
sawtooth-sabre-tp  | 2020-01-29 21:47:25,013 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] create_collection: 258
sawtooth-sabre-tp  | 2020-01-29 21:47:25,013 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] Attempting to get state, addresses: ["12cd3cde772f1b71c23958a18d88651ec5ce96354d445f9754f437711451ed5998033e"]
sawtooth-sabre-tp  | 2020-01-29 21:47:25,035 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] GET_STATE Execution time: 0 secs 22 ms
sawtooth-sabre-tp  | 2020-01-29 21:47:25,036 INFO  [sawtooth_sabre::wasm_executor::wasm_externals] Retrieving collection length. Head pointer 328
sawtooth-sabre-tp  | 2020-01-29 21:47:25,037 INFO  [sawtooth_sdk::processor] TP_PROCESS_REQUEST sending TpProcessResponse: ExternalsError { message: "Trap(Trap { kind: Unreachable })" }
sawtooth-sabre-tp  | 2020-01-29 21:47:25,970 INFO  [sawtooth_sdk::processor] Message: 3c97c91e85234144ada8b2c9247f17e1
sawtooth-sabre-tp  | 2020-01-29 21:47:25,970 INFO  [sawtooth_sabre::handler] Action: Execute Contract ["00ec0152bd2cc34da20f708540ae225d736c3e187f06687f5c1315b7f5d312ab0b3f82", "00ec02f59c1970c55fbfbe90ec0958e83d7757a01887a651d12f63761f596215420f76", "00ec000c03751f946450dd3c4cf4adc580836d0b5334b61a79ad3e027ecb9678fda376", "12cd3c"] ["00ec0152bd2cc34da20f708540ae225d736c3e187f06687f5c1315b7f5d312ab0b3f82", "00ec02f59c1970c55fbfbe90ec0958e83d7757a01887a651d12f63761f596215420f76", "00ec000c03751f946450dd3c4cf4adc580836d0b5334b61a79ad3e027ecb9678fda376", "12cd3c"]```
In an infinite loop.

**Comment by RyanLassigBanks** [comment]
> Assuming we need a follow-on task to implement better error handling. (see my above comment).

Currently AssemblyScript doesn't support catching errors, but for some reason supports throwing exceptions. I was experimenting with adding Results to the sdk to alleviate this issue, but it will still throw a `trap` for exceptions thrown by lower level functions, like indexing into an array incorrectly.

Created [issue #1838](https://zenhub.cglcloud.com/app/workspaces/product-core-5d37298b0add7306b30efde2/issues/digitalsupplychain/product-core/1838) for tracking.

File: sdks/assemblyscript/assembly/env.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2018 Cargill Incorporated
```
Comment (agunde406) [inline sdks/assemblyscript/assembly/env.ts]: this could probably be updated to  Copyright 2018-2020 Cargill Incorporated

File: .gitignore [hunk 2]
```diff
@@ -14,8 +14,11 @@
 /integration/target/
 /integration/Cargo.lock
 
-/sdk/Cargo.lock
-/sdk/target/
+/sdks/rust/Cargo.lock
```
Comment (agunde406) [inline .gitignore]: This change should be mentioned in the commit message or separated out into its own commit

File: contracts/inc-dec-key/assembly/index.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    const address = computeAddress(name);
+    info("Creating address " + address);
+    const init_value = Uint8Array.wrap(String.UTF8.encode((0).toString()));
+
+    const array = new Array<StateEntry>(0);
+    array.push(new StateEntry(address, init_value));
+
+    ctx.setStateEntries(array);
+
+    info("Couter: " + name + " initialized!");
```
Comment (agunde406) [inline contracts/inc-dec-key/assembly/index.ts]: Couter -> Counter

File: contracts/inc-dec-key/manifest.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2018 Cargill Incorporated
```
Comment (agunde406) [inline contracts/inc-dec-key/manifest.yaml]: # Copyright 2018-2020 Cargill Incorporated

**Pull request merged**